File: FinalProject/FinalProject3/FinalProject3/FinalProject3/Models/Interaction.cs
using System.ComponentModel.DataAnnotations;

namespace FinalProject3.Models
{
    public class Interaction
    {
        [Key]
        public required string Id { get; set; }
        public string Link { get; set; } = string.Empty;
        public string ImageURL { get; set; } = string.Empty;
        [Required, MinLength(2)]
        public string Text { get; set; } = string.Empty;
        [Required]
        public required User Author { get; set; }

        public List<Votes> Votes { get; set; } = [];

        public int UpVotes { get; set; }

        public int DownVotes { get; set; }

        public int TotalVotes { get; set; }
    }
}

File: FinalProject/FinalProjectFront/src/Models/Interaction.ts
import { ISocialGroup } from "./SocialGroup";
import { IAppUserDisplay } from "./UserModels";

interface IInteraction {
  id: string;
  link: string;
  imageUrl: string;
  text: string;
  author: IAppUserDisplay;
  votes: Vote[];
  upVotes: number;
  downVotes: number;
  totalVotes: number;
  datetime: string;
  comments: IComment[];
  calcVotes: () => void;
}

interface Vote {
  id: string;
  voter: IAppUserDisplay;
  voted: number;
}

interface IPost extends IInteraction {
  title: string;
  category?: ICategory;
  group?: ISocialGroup;
  keyWords: string[];
}

interface IComment extends IInteraction {
  parentPost?: IPost;
  parentComment?: IComment;
}

interface ICategory {
  id: number;
  name: string;
}

export type { IInteraction, Vote, IPost, IComment, ICategory };
